src/main/java/frc/robot/subsystems/drive/DriveIOCTRE.java

```diff
@@ -8,41 +8,98 @@
+ * class to provide swerve drive functionality with additional telemetry and simulation support.
+ *
+ * <p>This class handles: - Odometry data collection and buffering - Vision measurement integration
+ * - Simulation state updates - Thread-safe telemetry updates
  */
 public class DriveIOCTRE extends SwerveDrivetrain implements DriveIO {
-  private static final double kSimLoopPeriod = 0.005; // 5 ms
-  private Notifier m_simNotifier;
-  private double m_lastSimTime;
+  // Simulation constants
+  private static final double SIMULATION_LOOP_PERIOD = 0.005; // 5 ms
+  private Notifier simulationNotifier;
+  private double lastSimulationTime;
 
+  // Queue configuration
+  private static final int QUEUE_SIZE = 20;
+
+  // Thread-safe data storage
+  private final Lock odometryLock = new ReentrantLock();
+  private final Queue<Rotation2d> gyroYawQueue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+  private final Queue<Double> timestampQueue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+  private final List<Queue<Double>> drivePositionQueues = new ArrayList<>();
+  private final List<Queue<Rotation2d>> steerPositionQueues = new ArrayList<>();
+
+  /**
+   * Creates a new DriveIOCTRE with specified update frequency.
+   *
+   * @param driveTrainConstants Constants defining drivetrain behavior
+   * @param odometryUpdateFrequency How often to update odometry in Hz
+   * @param modules Array of constants for each swerve module
+   */
   public DriveIOCTRE(
       SwerveDrivetrainConstants driveTrainConstants,
-      double OdometryUpdateFrequency,
+      double odometryUpdateFrequency,
       SwerveModuleConstants... modules) {
-    super(driveTrainConstants, OdometryUpdateFrequency, modules);
-
-    if (Utils.isSimulation()) {
-      startSimThread();
-    }
+    super(driveTrainConstants, odometryUpdateFrequency, modules);
+    initializeQueues();
+    // This pulls data from our Odometry thread or in this case at 250 Hz
+    registerTelemetry(this::updateTelemetry);
+    setupSimulation();
   }
 
+  /**
+   * Creates a new DriveIOCTRE with default update frequency.
+   *
+   * @param driveTrainConstants Constants defining drivetrain behavior
+   * @param modules Array of constants for each swerve module
+   */
   public DriveIOCTRE(
       SwerveDrivetrainConstants driveTrainConstants, SwerveModuleConstants... modules) {
     super(driveTrainConstants, modules);
+    initializeQueues();
+    // This pulls data from our Odometry thread or in this case at 250 Hz
+    registerTelemetry(this::updateTelemetry);
+    setupSimulation();
+  }
 
-    if (Utils.isSimulation()) {
+  /** Initializes the position queues for drive and steer data. */
+  private void initializeQueues() {
+    for (int i = 0; i < 4; i++) {
+      drivePositionQueues.add(new ArrayBlockingQueue<>(QUEUE_SIZE));
+      steerPositionQueues.add(new ArrayBlockingQueue<>(QUEUE_SIZE));
+    }
+  }
+
+  /** Sets up telemetry and simulation if needed. */
+  private void setupSimulation() {
```

Comment:
why woudl you call setupSimulation if the current mode wasn't SIM?

Reply:
can't hurt to have the redundant chekc though ;)

Reply:
We have a slight issue with SIM right now. There may need to be stuff added inside of that. Figured I would extract it so then I am not changing things in two locations

Reply:
> why woudl you call setupSimulation if the current mode wasn't SIM?

There's not a ton of easy ways of doing it through command based where you only call code during sim unfortunately. (simulationInit/simulationPeriodic in Robot.java)